yesod-hspec/test/main.hs

```diff
@@ -0,0 +1,662 @@
+                    setMethod "POST"
+                    setUrl ("check-hobby" :: Text)
+                    byLabelPrefix "hobby" "fishing"
+                res <- maybe "Couldn't get response" simpleBody <$> getResponse
+                assertEq "hobby isn't set" res "fishing"
+            yit "byLabelPrefix throws an error if it finds multiple labels" $ do
+                (bad :: Either SomeException ()) <- try (request $ do
+                    setMethod "POST"
+                    setUrl ("label-prefix-error" :: Text)
+                    byLabelPrefix "hobby" "fishing")
+                assertEq "failure wasn't called" (isLeft bad) True
+            yit "byLabelSuffix matches over the suffix of the labels" $ do
+                get ("/label-suffix" :: Text)
+                request $ do
+                    setMethod "POST"
+                    setUrl ("check-hobby" :: Text)
+                    byLabelSuffix "hobby" "fishing"
+                res <- maybe "Couldn't get response" simpleBody <$> getResponse
+                assertEq "hobby isn't set" res "fishing"
+            yit "byLabelSuffix throws an error if it finds multiple labels" $ do
+                (bad :: Either SomeException ()) <- try (request $ do
+                    setMethod "POST"
+                    setUrl ("label-suffix-error" :: Text)
+                    byLabelSuffix "hobby" "fishing")
+                assertEq "failure wasn't called" (isLeft bad) True
+        ydescribe "Content-Type handling" $ do
+            yit "can set a content-type" $ do
+                request $ do
+                    setUrl ("/checkContentType" :: Text)
+                    addRequestHeader ("Expected-Content-Type","text/plain")
+                    addRequestHeader ("Content-Type","text/plain")
+                statusIs 200
+            yit "adds the form-urlencoded Content-Type if you add parameters" $ do
+                request $ do
+                    setUrl ("/checkContentType" :: Text)
+                    addRequestHeader ("Expected-Content-Type","application/x-www-form-urlencoded")
+                    addPostParam "foo" "foobarbaz"
+                statusIs 200
+            yit "defaults to no Content-Type" $ do
+                get ("/checkContentType" :: Text)
+                statusIs 200
+            yit "returns a 415 for the wrong Content-Type" $ do
+                -- Tests that the test handler is functioning
+                request $ do
+                    setUrl ("/checkContentType" :: Text)
+                    addRequestHeader ("Expected-Content-Type","application/x-www-form-urlencoded")
+                    addRequestHeader ("Content-Type","text/plain")
+                statusIs 415
+
+    describe "hooks" $ yesodSpec app $ do
+        ybefore_ (get ("/" :: Text)) $ do
+            yit "works" $ do
+                statusIs 200
+            ybefore_ (post ("/cookie/foo" :: Text)) $ do
+                yit "works again" $ do
+                    statusIs 303
```

Comment:
This is the exciting thing. Now if you want a bunch of tests that share some setup, you can get it. This makes it much easier to write small tests that make small assertions and reduce code duplication.